oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/ThrottlingDocumentStoreWrapper.java

```diff
@@ -0,0 +1,241 @@
+        return store.invalidateCache(keys);
+    }
+
+    @Override
+    public <T extends Document> void invalidateCache(Collection<T> collection, String key) {
+        store.invalidateCache(collection, key);
+    }
+
+    @Override
+    public void dispose() {
+        store.dispose();
+    }
+
+    @Override
+    public <T extends Document> T getIfCached(final Collection<T> collection, final String key) {
+        return store.getIfCached(collection, key);
+    }
+
+    @Override
+    public void setReadWriteMode(String readWriteMode) {
+        store.setReadWriteMode(readWriteMode);
+    }
+
+    @Override
+    public Iterable<CacheStats> getCacheStats() {
+        return store.getCacheStats();
+    }
+
+    @Override
+    public Map<String, String> getMetadata() {
+        return store.getMetadata();
+    }
+
+    @NotNull
+    @Override
+    public Map<String, String> getStats() {
+        return store.getStats();
+    }
+
+    @Override
+    public long determineServerTimeDifferenceMillis() {
+        return store.determineServerTimeDifferenceMillis();
+    }
+
+    /**
+     * Return the size limit for node name based on the document store implementation
+     *
+     * @return node name size limit
+     */
+    @Override
+    public int getNodeNameLimit() {
+        return store.getNodeNameLimit();
+    }
+
+    /**
+     * Return the @{@link ThrottlingMetrics} for the underlying document store
+     *
+     * @return throttling metric for document store
+     */
+    @Override
+    public ThrottlingMetrics throttlingMetrics() {
+        return store.throttlingMetrics();
+    }
+
+    // helper methods
+
+    private void performThrottling() {
+
+        final ThrottlingMetrics metrics = throttlingMetrics();
+        long throttleTime = getThrottleTime(metrics);
+
+        if (throttleTime == 0) {
+            return; // no throttling
+        }
+
+        try {
+            LOG.info("Throttling the system for {} ms", throttleTime);
+            sleep(throttleTime);
+        } catch (InterruptedException e) {
+            // swallow the exception and log it
+            LOG.error("Error while throttling", e);
+        }
+    }
+
+     @VisibleForTesting
+     long getThrottleTime(final ThrottlingMetrics metrics) {
+
+        long throttleTime = metrics.throttlingTime();
+        final double threshold = metrics.threshold();
+        final double currValue = metrics.currValue();
+
+        if (DoubleMath.fuzzyCompare(currValue,threshold/8,  0.001) <= 0) {
+            throttleTime = throttleTime * 8;
+        } else if (DoubleMath.fuzzyCompare(currValue,threshold/4, 0.001) <= 0) {
+            throttleTime = throttleTime * 4;
+        } else if (DoubleMath.fuzzyCompare(currValue, threshold/2, 0.001) <= 0) {
+            throttleTime = throttleTime * 2;
+        } else if (DoubleMath.fuzzyCompare(currValue, threshold,0.001) <= 0) {
+            throttleTime = metrics.throttlingTime();
```

Comment:
I don't understand this logic. Why is throttleTime only multiplied around certain thresholds and not when the value is between thresholds?

Reply:
@mreutegg I am updating the throttling time in multiple of 2 after in conjunction with every drop (in multiples of 2) in the current value as compared to the threshold.

Reply:
OK, I see. Somehow I was confused by the threshold with the comparison. What is the reason for not using `Double.compare()` instead?

Reply:
> Double.compare() doesn't take tolerance into account while comparing 2 doubles

But why is this needed here?

Reply:
I agree that we can go with `Double.compare()` but, then we would only be throttling after reaching the threshold, any value which is very close to the threshold value would be ignored, so I wanted to handle those cases as well.

P.S. Even `Double.compare()` would work just fine here.

Reply:
OK. I still think `Double.compare()` is sufficient, but I don't mind using `DoubleMath.fuzzyCompare()`.